src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs

```diff
@@ -717,121 +479,33 @@ private IAsyncResult BeginRead(AsyncCallback callback, Object state)
                 // ReadFile when using overlapped structures!  This is by design NT behavior.
                 if (r == -1)
                 {
-                    // One side has closed its handle or server disconnected. Set the state to Broken 
-                    // and do some cleanup work
-                    if (errorCode == Interop.mincore.Errors.ERROR_BROKEN_PIPE ||
-                        errorCode == Interop.mincore.Errors.ERROR_PIPE_NOT_CONNECTED)
+                    switch (errorCode)
                     {
-                        State = PipeState.Broken;
-
-                        // Clear the overlapped status bit for this special case. Failure to do so looks 
-                        // like we are freeing a pending overlapped.
-                        intOverlapped->InternalLow = IntPtr.Zero;
-
-                        // EndRead will free the Overlapped struct
-                        asyncResult.CallUserCallback();
-                    }
-                    else if (errorCode != Interop.mincore.Errors.ERROR_IO_PENDING)
-                    {
-                        throw Win32Marshal.GetExceptionForWin32Error(errorCode);
-                    }
-                }
-                ReadWriteAsyncParams readWriteParams = state as ReadWriteAsyncParams;
-                if (readWriteParams != null)
-                {
-                    if (readWriteParams.CancellationHelper != null)
-                    {
-                        readWriteParams.CancellationHelper.AllowCancellation(_handle, intOverlapped);
+                        // One side has closed its handle or server disconnected.
+                        // Set the state to Broken and do some cleanup work
+                        case Interop.mincore.Errors.ERROR_BROKEN_PIPE:
+                        case Interop.mincore.Errors.ERROR_PIPE_NOT_CONNECTED:
+                            State = PipeState.Broken;
+
+                            // Clear the overlapped status bit for this special case. Failure to do so looks 
+                            // like we are freeing a pending overlapped.
+                            completionSource.Overlapped->InternalLow = IntPtr.Zero;
+                            completionSource.ReleaseResources();
+                            completionSource.SetCompletedSynchronously();
+                            UpdateMessageCompletion(true);
```

Comment:
TODO: remove this line because `SetCompletedSynchronously` also invokes `UpdateMessageCompletion`